src/engine/pieces/pawn.ts

```diff
@@ -1,13 +1,42 @@
 import Piece from './piece';
 import Player from '../player';
 import Board from '../board';
+import Square from '../square';
 
 export default class Pawn extends Piece {
     public constructor(player: Player) {
         super(player);
     }
 
     public getAvailableMoves(board: Board) {
-        return new Array(0);
+        let availableMoves = new Array<Square>();
+        let finalAvailableMoves = [];
+        let currentPosition = board.findPiece(this);
+        let oldRow = currentPosition.row;
+        let oldCol = currentPosition.col;
+
+        if (this.player == 0) {
+            this.addAvailableMoveAndContinue(availableMoves, oldRow+1, oldCol, currentPosition, board);
+            if (oldRow == 1 && !board.getPiece(new Square(oldRow+1, oldCol))) {
+                this.addAvailableMoveAndContinue(availableMoves, oldRow+2, oldCol, currentPosition, board);
+            }
+        }
+        
+        else {
+            this.addAvailableMoveAndContinue(availableMoves, oldRow-1, oldCol, currentPosition, board);
+            if (oldRow == Piece.boardSize-2 && !board.getPiece(new Square(oldRow-1, oldCol))) {
+                this.addAvailableMoveAndContinue(availableMoves, oldRow-2, oldCol, currentPosition, board);
+            }
+        }
+
+        for (let index = 0; index < availableMoves.length; index++) {
```

Comment:
would be nice for this to be split up into nicely named functions so it's clearer what this is doing

Reply:
e.g. filterOutOccupiedSpaces

Reply:
And this can be shortened by filtering moves out of `availableMoves` rather than creating a new list (this is also good for memory saving) by
e.g. `availableMoves = availableMoves.filter(pos => !board.getPiece(pos))`

Reply:
added filter. Left in the blocks of code in lines 19-22 & 26-29 as we felt that we shouldn't create a new function just for 2 blocks of code. (let us know if this is not best practice)